File: sdk-java/src/test/java/io/superdurable/dex/integ/PersistenceTest.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Sustainable Use License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Client;
import io.superdurable.dex.Attribute;
import io.superdurable.dex.AttributeMatch;
import io.superdurable.dex.StartFlowOptions;
import io.superdurable.dex.WaitForAttributeOptions;
import io.superdurable.dex.exceptions.RequestTimeoutException;
import io.superdurable.dex.testing.DexDevTestEnvironment;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

@Tag("dex-dev")
public final class PersistenceTest {
    private static final PersistenceWorkflow WORKFLOW = new PersistenceWorkflow();
    private static final PersistenceSetAttributesWorkflow SET_ATTRIBUTES_WORKFLOW =
            new PersistenceSetAttributesWorkflow();

    @TempDir
    Path cacheDirectory;

    @Test
    void testPersistenceReads() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = "persistence-" + UUID.randomUUID();
            final StartFlowOptions options = StartFlowOptions.newBuilder()
                    .addAttribute(WORKFLOW.initial, "initial")
                    .addAttribute(WORKFLOW.dataMap, "one", "initial")
                    .build();
            environment.client().startFlow(WORKFLOW, flowId, "input", options);
            assertEquals("input", environment.client().waitForFlow(flowId, Duration.ofSeconds(30)).getSingleOutput(String.class));
        }
    }

    @Test
    void testSetIndexedAttributes() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                SET_ATTRIBUTES_WORKFLOW)) {
            final String flowId = "set-indexed-attributes-" + UUID.randomUUID();
            environment.client().startFlow(SET_ATTRIBUTES_WORKFLOW, flowId, "start");
            final PersistenceSetAttributesWorkflow stub = environment.client().newRpcStub(
                    PersistenceSetAttributesWorkflow.class,
                    flowId);
            environment.client().invokeRPC(stub::setIndexed);
            environment.client().invokeRPC(stub::complete);

            assertEquals("test-result", environment.client().waitForFlow(flowId, Duration.ofSeconds(30)).getSingleOutput(String.class));
        }
    }

    @Test
    void testSetDataAttributes() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                SET_ATTRIBUTES_WORKFLOW)) {
            final String flowId = "set-data-attributes-" + UUID.randomUUID();
            final PersistenceWorkflow.ModelInput model = new PersistenceWorkflow.ModelInput();
            model.value = 7;
            environment.client().startFlow(SET_ATTRIBUTES_WORKFLOW, flowId, "start");
            final PersistenceSetAttributesWorkflow stub = environment.client().newRpcStub(
                    PersistenceSetAttributesWorkflow.class,
                    flowId);
            assertThrows(
                    RequestTimeoutException.class,
                    () -> environment.client().waitForAttributeMatch(
                            flowId,
                            SET_ATTRIBUTES_WORKFLOW.data,
                            AttributeMatch.equalTo("never"),
                            waitOptions(flowId, "never", Duration.ofSeconds(1))));
            final CompletableFuture<String> waiting = CompletableFuture.supplyAsync(
                    () -> environment.client().waitForAttributeMatch(
                            flowId,
                            SET_ATTRIBUTES_WORKFLOW.data,
                            AttributeMatch.equalTo("query-start"),
                            waitOptions(flowId, "data", Duration.ofSeconds(30))));
            environment.client().invokeRPC(stub::setData, "query-start");
            assertEquals("query-start", waiting.get(30, TimeUnit.SECONDS));
            final CompletableFuture<String> waitingMap = CompletableFuture.supplyAsync(
                    () -> environment.client().waitForAttributeMatch(
                            flowId,
                            SET_ATTRIBUTES_WORKFLOW.dataMap,
                            "special % key",
                            AttributeMatch.equalTo("mapped-value"),
                            waitOptions(flowId, "map", Duration.ofSeconds(30))));
            environment.client().invokeRPC(stub::setMapOne, "mapped-value");
            environment.client().invokeRPC(stub::setMapSpecial, "mapped-value");
            assertEquals("mapped-value", waitingMap.get(30, TimeUnit.SECONDS));
            environment.client().invokeRPC(stub::setInteger, 3);
            assertEquals(
                    3,
                    environment.client().waitForAttributeMatch(
                            flowId,
                            SET_ATTRIBUTES_WORKFLOW.integer,
                            AttributeMatch.greaterThan(0),
                            waitOptions(flowId, "integer", Duration.ofSeconds(30))));
            assertThrows(
                    IllegalArgumentException.class,
                    () -> environment.client().waitForAttributeMatch(
                            flowId,
                            SET_ATTRIBUTES_WORKFLOW.model,
                            AttributeMatch.equalTo(model),
                            waitOptions(flowId, "model", Duration.ofSeconds(30))));
            assertThrows(
                    IllegalArgumentException.class,
                    () -> environment.client().waitForAttributeMatch(
                            flowId,
                            Attribute.define("bytes", byte[].class),
                            AttributeMatch.equalTo(new byte[] {1}),
                            waitOptions(flowId, "bytes", Duration.ofSeconds(30))));
            assertThrows(
                    IllegalArgumentException.class,
                    () -> environment.client().waitForAttributeMatch(
                            flowId,
                            Attribute.define("null", Void.class),
                            AttributeMatch.equalTo((Void) null),
                            waitOptions(flowId, "null", Duration.ofSeconds(30))));
            environment.client().invokeRPC(stub::setModel, model);
            environment.client().invokeRPC(stub::complete);

            assertEquals("test-result", environment.client().waitForFlow(flowId, Duration.ofSeconds(30)).getSingleOutput(String.class));
        }
    }

    void compilePersistenceReads(final Client client) {
        final StartFlowOptions options = StartFlowOptions.newBuilder()
                .addAttribute(WORKFLOW.initial, "initial")
                .addAttribute(WORKFLOW.dataMap, "one", "initial")
                .build();
        client.startFlow(WORKFLOW, "persistence", "input", options);
        final String output = client.waitForFlow("persistence").getSingleOutput(String.class);
        consume(output);
    }

    void compilePersistenceWrites(final Client client) {
        client.startFlow(SET_ATTRIBUTES_WORKFLOW, "set-attributes", "input");
        final PersistenceSetAttributesWorkflow stub = client.newRpcStub(
                PersistenceSetAttributesWorkflow.class,
                "set-attributes");
        client.invokeRPC(stub::setData, "value");
        client.invokeRPC(stub::setMapOne, "value");
        client.invokeRPC(stub::setIndexed);
        final String matched = client.waitForAttributeMatch(
                "set-attributes",
                SET_ATTRIBUTES_WORKFLOW.data,
                AttributeMatch.equalTo("value"),
                waitOptions("set-attributes", "data", Duration.ofSeconds(30)));
        client.invokeRPC(stub::complete);
        final String matchedMap = client.waitForAttributeMatch(
                "set-attributes",
                SET_ATTRIBUTES_WORKFLOW.dataMap,
                "one",
                AttributeMatch.equalTo("value"),
                waitOptions("set-attributes", "map", Duration.ofSeconds(30)));
        final String output = client.waitForFlow("set-attributes").getSingleOutput(String.class);
        consume(matched, matchedMap, output);
    }

    private static void consume(final Object... values) {
    }

    private static WaitForAttributeOptions waitOptions(
            final String flowId,
            final String suffix,
            final Duration requestTimeout) {
        return WaitForAttributeOptions.newBuilder()
                .requestId(flowId + "-wait-" + suffix)
                .requestTimeout(requestTimeout)
                .build();
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/persistence_test.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Sustainable Use License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::time::Duration;

use dex_sdk::{Attribute, AttributeMatch, Client, Registry, SdkError, SdkResult, StartFlowOptions};

use crate::persistence_set_attributes_workflow::{
    self as set_attributes, PersistenceSetAttributesWorkflow,
};
use crate::persistence_workflow::{self as persistence, PersistenceModel, PersistenceWorkflow};
use crate::support::{DexDevTestEnvironment, flow_id};

#[test]
#[ignore = "requires dexcli dev"]
fn test_persistence_reads() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(PersistenceWorkflow::new()));
    let workflow = PersistenceWorkflow::new();
    let flow_id = flow_id("persistence");
    let options = StartFlowOptions::new()
        .initial_attribute(&persistence::INITIAL, "initial".to_string())
        .initial_attribute_map(&workflow.data_map, "one", "initial".to_string());
    environment
        .client
        .start_flow_with_options(&workflow, &flow_id, "input".to_string(), options)
        .expect("start persistence Flow");
    assert_eq!(
        "input",
        environment
            .client
            .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(30))
            .and_then(|result| result.single_output::<String>())
            .expect("complete persistence Flow")
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_set_indexed_attributes() {
    let environment = DexDevTestEnvironment::start(
        Registry::new().register(PersistenceSetAttributesWorkflow::new()),
    );
    let workflow = PersistenceSetAttributesWorkflow::new();
    let flow_id = flow_id("set-indexed-attributes");
    environment
        .client
        .start_flow(&workflow, &flow_id, "start".to_string())
        .expect("start set-indexed-attributes Flow");
    environment
        .client
        .invoke_rpc_without_input::<()>(&flow_id, PersistenceSetAttributesWorkflow::SET_INDEXED)
        .expect("set indexed Attributes through RPC");
    environment
        .client
        .invoke_rpc_without_input::<()>(&flow_id, PersistenceSetAttributesWorkflow::COMPLETE)
        .expect("complete through RPC");
    assert_eq!(
        "test-result",
        environment
            .client
            .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(30))
            .and_then(|result| result.single_output::<String>())
            .expect("complete set-indexed-attributes Flow")
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_set_data_attributes() {
    let environment = DexDevTestEnvironment::start(
        Registry::new().register(PersistenceSetAttributesWorkflow::new()),
    );
    let workflow = PersistenceSetAttributesWorkflow::new();
    let flow_id = flow_id("set-data-attributes");
    environment
        .client
        .start_flow(&workflow, &flow_id, "start".to_string())
        .expect("start set-data-attributes Flow");
    assert!(matches!(
        environment.client.wait_for_attribute_match(
            &flow_id,
            &set_attributes::DATA,
            AttributeMatch::equal_to("never".to_string()),
            Duration::from_secs(1),
        ),
        Err(SdkError::LongPollTimeout { .. })
    ));
    std::thread::scope(|scope| {
        let waiting = scope.spawn(|| {
            environment.client.wait_for_attribute_match(
                &flow_id,
                &set_attributes::DATA,
                AttributeMatch::equal_to("query-start".to_string()),
                Duration::from_secs(30),
            )
        });
        environment
            .client
            .invoke_rpc(
                &flow_id,
                PersistenceSetAttributesWorkflow::SET_DATA,
                "query-start".to_string(),
            )
            .expect("set data Attribute");
        let matched = waiting
            .join()
            .expect("join Attribute wait")
            .expect("wait for data Attribute");
        assert_eq!("query-start", matched);
    });
    std::thread::scope(|scope| {
        let waiting = scope.spawn(|| {
            environment.client.wait_for_attribute_map_instance_match(
                &flow_id,
                &set_attributes::DATA_MAP,
                "special % key",
                AttributeMatch::equal_to("mapped-value".to_string()),
                Duration::from_secs(30),
            )
        });
        environment
            .client
            .invoke_rpc(
                &flow_id,
                PersistenceSetAttributesWorkflow::SET_MAP_SPECIAL,
                "mapped-value".to_string(),
            )
            .expect("set special AttributeMap entry");
        let matched = waiting
            .join()
            .expect("join AttributeMap wait")
            .expect("wait for AttributeMap entry");
        assert_eq!("mapped-value", matched);
    });
    environment
        .client
        .invoke_rpc(&flow_id, PersistenceSetAttributesWorkflow::SET_INTEGER, 3)
        .expect("set revision Attribute");
    assert_eq!(
        3,
        environment
            .client
            .wait_for_attribute_match(
                &flow_id,
                &set_attributes::INTEGER,
                AttributeMatch::greater_than(0),
                Duration::from_secs(30),
            )
            .expect("wait for revision Attribute")
    );
    assert!(matches!(
        environment.client.wait_for_attribute_match(
            &flow_id,
            &set_attributes::MODEL,
            AttributeMatch::equal_to(PersistenceModel { value: 8 }),
            Duration::from_secs(30),
        ),
        Err(SdkError::InvalidArgument { .. })
    ));
    assert!(matches!(
        environment.client.wait_for_attribute_match(
            &flow_id,
            &Attribute::<Vec<u8>>::new("bytes"),
            AttributeMatch::equal_to(vec![1]),
            Duration::from_secs(30),
        ),
        Err(SdkError::InvalidArgument { .. })
    ));
    assert!(matches!(
        environment.client.wait_for_attribute_match(
            &flow_id,
            &Attribute::<()>::new("null"),
            AttributeMatch::equal_to(()),
            Duration::from_secs(30),
        ),
        Err(SdkError::InvalidArgument { .. })
    ));
    environment
        .client
        .invoke_rpc(
            &flow_id,
            PersistenceSetAttributesWorkflow::SET_MAP_ONE,
            "mapped-value".to_string(),
        )
        .expect("set AttributeMap entry");
    environment
        .client
        .invoke_rpc(
            &flow_id,
            PersistenceSetAttributesWorkflow::SET_MODEL,
            PersistenceModel { value: 7 },
        )
        .expect("set model Attribute");
    environment
        .client
        .invoke_rpc_without_input::<()>(&flow_id, PersistenceSetAttributesWorkflow::COMPLETE)
        .expect("complete through RPC");
    assert_eq!(
        "test-result",
        environment
            .client
            .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(30))
            .and_then(|result| result.single_output::<String>())
            .expect("complete set-data-attributes Flow")
    );
}

#[allow(dead_code)]
fn compile_persistence_reads(client: &Client) -> SdkResult<()> {
    let workflow = PersistenceWorkflow::new();
    let options = StartFlowOptions::new()
        .initial_attribute(&persistence::INITIAL, "initial".to_string())
        .initial_attribute_map(&workflow.data_map, "one", "initial".to_string());
    client.start_flow_with_options(&workflow, "persistence", "input".to_string(), options)?;
    let _: String = client.wait_for_flow("persistence")?.single_output()?;
    Ok(())
}

#[allow(dead_code)]
fn compile_persistence_writes(client: &Client) -> SdkResult<()> {
    let workflow = PersistenceSetAttributesWorkflow::new();
    client.start_flow(&workflow, "set-attributes", "input".to_string())?;
    client.invoke_rpc(
        "set-attributes",
        PersistenceSetAttributesWorkflow::SET_DATA,
        "value".to_string(),
    )?;
    client.invoke_rpc(
        "set-attributes",
        PersistenceSetAttributesWorkflow::SET_MAP_ONE,
        "value".to_string(),
    )?;
    client.invoke_rpc_without_input::<()>(
        "set-attributes",
        PersistenceSetAttributesWorkflow::SET_INDEXED,
    )?;
    let _: String = client.wait_for_attribute_match(
        "set-attributes",
        &set_attributes::DATA,
        AttributeMatch::equal_to("value".to_string()),
        Duration::from_secs(30),
    )?;
    let _: String = client.wait_for_attribute_map_instance_match(
        "set-attributes",
        &set_attributes::DATA_MAP,
        "one",
        AttributeMatch::equal_to("value".to_string()),
        Duration::from_secs(30),
    )?;
    client.invoke_rpc_without_input::<()>(
        "set-attributes",
        PersistenceSetAttributesWorkflow::COMPLETE,
    )?;
    let _: String = client.wait_for_flow("set-attributes")?.single_output()?;
    Ok(())
}
